### PR TITLE
Add Generative Modelling by Estimating Gradients of Data Distribution…

### PR DESCRIPTION
**Pull request by dvmazur**
I've implemented a tutorial based on a recent paper [__Generative Modeling by Estimating Gradients of the Data Distribution__](https://arxiv.org/abs/1907.05600).

This tutorial showcases JAX advanced autograd capabilities (e.g. Jacobian-Vector products, `jax.jvp`) in order to implement the math behind Sliced Score Matching.

In order to view the actual notebook, [click here](https://nbviewer.jupyter.org/gist/deniskamazur/2ea41beeab7895166393f74ddd16235f) or [Run in Colab](https://colab.research.google.com/gist/deniskamazur/2ea41beeab7895166393f74ddd16235f/score_matching_jax.ipynb). The code in this PR contains urls to colab that will only work once merged into master.

I would also appreciate a feedback on how i can  make this tutorial easier to follow.

### CODE REVIEW

**Comment by mattjj** [comment]
Thanks so much for making and sharing this! It looks fantastic. I learned a lot from reading it, and it's a great illustration of JAX's autodiff, especially forward mode. Beautiful work!

It looks good to me to merge now, but here are a few minor suggestions:

1. You could add a comment explaining why it's a good idea to use `jacfwd` in the first (full-Jacobian-computing) example, namely that (IIUC) the dimension of the data input for `net_apply` is only 2, and that's the same as the output dimension. Do we always expect that to be the case? (I'm actually not sure, but it seems plausible since we're learning a model mapping `x` to an approximation of `∇ log p(x)`! It might be worth commenting on that since it underscores why forward-mode autodiff is valuable here.)
2. I'm not sure if they're worth linking, but Yee Whye Teh and his students/coauthors have some follow-up work on the properties of SGLD (https://arxiv.org/abs/1409.0578 and https://arxiv.org/abs/1501.00438) discussing notions of convergence as \epsilon \to 0.
3. This is mostly just my personal curiosity now: in the Sliced Score Matching section, could we sample v to be ±1 valued (i.e. Rademacher variables) rather than Gaussian to reduce the variance? I think that's a trick that can be useful when estimating traces in this way. 

Okay, other than the first suggestion there, I mainly wanted to share those thoughts to show that I'd read your work thoroughly :) 

Let me know what you think about adding a comment on dimensionality and how it motivates the use of forward-mode, then let's merge!

**Comment by deniskamazur** [comment]
Thanks for your suggestions! 

1. Neat idea. I've added it as a side note in the last commit.
2. I've added this as further reading in the Langevin dynamics section.
3. You can indeed use Rademacher distribution to estimate the trace. In their Pytorch implementation, authors support both Gaussian and Rademacher vectors. I used gaussian noise for brevity.

**Review by mattjj — APPROVED** [review]
LGTM! Thanks again for this beautiful notebook.
